Printing with custom page size and reading advanced printer options

alexeyst opened the pull request:
This commit adds support for retrieving printer supported media sizes as well as tons of other information via CUPS; it is made available through `getPrinter(<name>).driver_options`; furthermore, a user can now define page size via printDirect `media` option as well as manipulate scaling via `fit_to_page`.

Tested on OSX, not tested on Windows (CUPS is only for posix, but input parameters are passed to Windows too).

That was just a quick fix for my own project that I thought I'd share.

pushpak commented:
:+1:

tojocky commented on src/node_printer_posix.cc:
```diff
@@ -132,6 +163,30 @@ namespace
+        }
+        result_printer->Set(V8_STRING_NEW_UTF8("driver_options"), ppd_options);
+
         v8::Local<v8::Object> result_printer_options = V8_VALUE_NEW_DEFAULT_V_0_11_10(Object);
```
This information may not be required for most of the time. I think this would be better to be in a separate routine, let say, "getPrinterDriverOptions". or to be activated by parameter.

tojocky commented on src/node_printer_posix.cc:
```diff
@@ -342,6 +397,9 @@ MY_NODE_MODULE_CALLBACK(PrintDirect)
     REQUIRE_ARGUMENT_STRING(iArgs, 3, type);
+    REQUIRE_ARGUMENT_STRING(iArgs, 4, media);
+    REQUIRE_ARGUMENT_STRING(iArgs, 5, fit_to_page);
+
```
Could you please convert into an Array?

alexeyst commented:
OK @tojocky, I'll try to make the changes as soon as I can, thanks for your comments! I might not get back very quickly as I've got a few other things on my plate for the upcoming week, and I'm not a C++ developer in the first place :)

tojocky commented:
Thanks for sharing, I will finish the code with windows support.

alexeyst commented:
Hey @tojocky!

Actually I've decided to go ahead and implement the changes right away not to keep you waiting. Here's what I've done:

Refactored driver_options retrieval into a separate method `getPrinterDriverOptions`
Made possible an unlimited supply of CUPS options via a JS object passed into printDirect
Modified README and one example accordingly
Fixed a little bug with getPrinter not returning a value (introduced by the correcting function for printing info)

tojocky commented on lib/printer.js:
```diff
@@ -1,8 +1,8 @@
-    path = require("path"),
+fs = require("fs")
+child_process = require("child_process")
+os = require("os")
```
missing commas

alexeyst replied:
I haven't modified that code, I am traveling now and not available to insert right away, will try doing that a little later.

tojocky commented on src/node_printer_posix.cc:
```diff
@@ -349,8 +445,21 @@ MY_NODE_MODULE_CALLBACK(PrintDirect)
+
+    v8::Local<v8::Array> props = print_options->GetPropertyNames();
+
+    for(unsigned int i = 0; i < props->Length(); i++) {
```
please prefix ++/-- operators. Suffix operators creates a temp value where here is useless.

alexeyst replied:
All right

tojocky commented:
This version looks much better. the only thing you brake windows build. I think you need to add body method of getPrinterDriverOptions to `node_printer_win.cc`

tojocky commented:
I will merge this PR and will ad extra changes. thanks for your PR.

alexeyst commented:
Right on @tojocky. I haven't modified `node_printer_win.cc` as you mentioned that you were planning to complete the Windows part on your own.

So, thanks for merging it in, and many thanks for this node module, it was a pleasure to be a (little) part of this.
